File: src/main/java/de/take_weiland/mods/commons/net/NetworkChannel.java
package de.take_weiland.mods.commons.net;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;

import java.util.function.Predicate;

/**
 * @author diesieben07
 */
public interface NetworkChannel<P> {

    void sendToServer(P packet);

    void sendTo(P packet, EntityPlayer player);

    void sendTo(P packet, Iterable<? extends EntityPlayer> players);

    void sendTo(P packet, Iterable<? extends EntityPlayer> players, Predicate<? super EntityPlayerMP> filter);

}

File: src/main/kotlin/de/takeweiland/mods/commons/net/base/NetworkChannel.kt
package de.takeweiland.mods.commons.net.base

import io.netty.buffer.ByteBuf
import net.minecraft.entity.player.EntityPlayer

/**
 * @author Take Weiland
 */
interface NetworkChannel<M : NetworkSerializable> {

    val name: String

    fun receive(buf: ByteBuf, player: EntityPlayer)

    fun receive(message: M, player: EntityPlayer)

    fun encode(message: M, buf: ByteBuf)

}